server/chat-plugins/announcements.ts

```diff
@@ -42,7 +53,47 @@ export class Announcement {
 	}
 
 	end() {
+		this.endTimer();
 		this.room.send(`|uhtmlchange|announcement${this.announcementNumber}|<div class="infobox">(${this.room.tr`The announcement has ended.`})</div>`);
+		delete this.room.settings.minorActivity;
+		this.room.minorActivity = null;
+		this.room.saveSettings();
+	}
+	save() {
+		const entry: AnnouncementData = {
+			source: this.source,
+			announcementNumber: this.announcementNumber,
+			timeoutMins: this.timeoutMins,
+			timerEnd: this.timerEnd,
+			activityId: 'announcement',
+		};
+		this.room.settings.minorActivity = entry;
+		this.room.saveSettings();
+		return entry;
+	}
+	runTimeout(timeout: number) {
+		this.timeoutMins = timeout;
+		this.timerEnd = Date.now() + timeout * 60000;
+		this.timeout = setTimeout(() => {
+			const room = this.room;
+			if (!room) return; // do nothing if the room doesn't exist anymore
+			if (room.minorActivity) room.minorActivity.end();
+		}, timeout * MINUTE);
+		this.save();
+		return this.timeout;
+	}
+	endTimer() {
+		if (!this.timeout) return;
+		clearTimeout(this.timeout);
+		this.timeoutMins = 0;
+		delete this.timerEnd;
+		return this;
+	}
+}
+
+for (const room of Rooms.rooms.values()) { // hotpatching!
+	if (room.settings.minorActivity?.activityId === 'announcement') {
+		room.minorActivity = new Announcement(room, room.settings.minorActivity);
```

Comment:
The saving seems entirely unnecessary? Why not just

```suggestion
	if (room.minorActivity?.activityId === 'announcement') {
		room.minorActivity = new Announcement(room, room.minorActivity.toJSON());
```

Reply:
(The same approach applies to hotpatching helptickets.)

Reply:
Oh, you're also trying to persist through restarts, aren't you? In which case this should stay as it is.

Reply:
Do you clear the old timer on hotpatch? It doesn't seem like you do.

Reply:
It does now.